### DISCUSSION TITLE
Lesson 12 depositCollateral() OSCEngine.sol

### DISCUSSION THREAD

**Discussion by kartik-giri** [Cyfrin/foundry-full-course-cu#1208] [Q&A] [answered]
I have a confusion in  `depositCollateral()` function we are calling `transferFrom()` function without giving allowance to the OSCEngine contract which is calling the function `transferFrom()`.
My question is should we not first give allowance and than call the transferFrom() function. Because here spender has no allowance to spend on the behalf of owner.

Here is code
```solidity
 function depositCollateral(address tokenCollateralAddress, uint256 amountCollateral)
        public
        moreThanZero(amountCollateral)
        nonReentrant
        isAllowedToken(tokenCollateralAddress)
    {
        s_collateralDeposited[msg.sender][tokenCollateralAddress] += amountCollateral;
        emit CollateralDeposited(msg.sender, tokenCollateralAddress, amountCollateral);
        bool success = IERC20(tokenCollateralAddress).transferFrom(msg.sender, address(this), amountCollateral);
        if (!success) {
            revert OSCEngine__TransferFailed();
        }}

**Comment by ghost** [comment, marked as answer]
The user must approve the engine as a spender in a separate transaction, then they can call `depositCollateral`. It's not cool if the app make the call to the approve method without asking for our permission.

**Reply by kartik-giri** [reply]
Thanks, I get it now.
